Document FX debugging

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50926 Document FX debugging**

Differential Revision: [D26068662](https://our.internmc.facebook.com/intern/diff/D26068662)

Commit:
Document FX debugging

[ghstack-poisoned]

Commit:
Update on "Document FX debugging"




[ghstack-poisoned]

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -293,10 +293,188 @@ Writing Transformations
+^^^^^^^^^^^^^^^^
+
+Once we've generated Python code from our symbolically-traced Module,
+we'll want to verify that the proper semantics were preserved. If they
```
I think we want to emphasize debugging transformations here. Debugging issues with symbolic tracing should (hopefully) be rare for external users

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -293,10 +293,188 @@ Writing Transformations
+to the generated code.
+
+If you’re not familiar with debuggers, please see the auxiliary section
+:ref:`Available Debuggers`.
```
This reference link doesn't work, I think you'll need to add an anchor tag like here: https://github.com/pytorch/pytorch/pull/50638/files#diff-8cd3ad665631e6f4e42aed8aa144620de5a93efa7e0388094ff3485c3334e0faR195
and refer to that

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -293,10 +293,188 @@ Writing Transformations
+
+Because FX generates the ``forward()`` function on GraphModules, using
+traditional debugging techniques like ``print`` statements or ``pdb`` is
+ not as straightfoward. Luckily, we have several techniques we can use
```
Delete leading space; this is causing a paragraph break in the rendered doc

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -293,10 +293,188 @@ Writing Transformations
+for debugging the generated code.
+
+``pdb``
+~~~~~~~~~~~~
```
The code backticks are making it so that this is not a heading, we can just have "pdb" as a raw stirng

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -293,10 +293,188 @@ Writing Transformations
+into it manually using ``pdb`` when the forward pass is invoked.
+
+::
+    x, y = torch.rand(2, 3), torch.rand(2, 3)
```
I think there needs to be an empty line before the code here to render properly

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -293,10 +293,188 @@ Writing Transformations
+    import pdb; pdb.set_trace()
+    m(x, y)
+
+Print the generated code
```
Capitalize this title

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -293,10 +293,188 @@ Writing Transformations
+    #       x = self.x
+    #       add_1 = x + y;  x = y = None
+    #       return add_1
+    # Copy this code for later
```
To show program outputs I prefer to use triple quotes
```
"""
   def forward(self, y):
       x = self.x
       add_1 = x + y;  x = y = None
       return add_1
"""
```

Or a separate code block in the RST

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -293,10 +293,188 @@ Writing Transformations
+    post_trace = SubclassM()
+
+
+Use the ``to_folder`` function from ``GraphModule``
```
Capitalize

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -293,10 +293,188 @@ Writing Transformations
+
+Use the ``to_folder`` function from ``GraphModule``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+``to_folder`` is a method in ``GraphModule`` that allows you to dump
```
Use :ref:`GraphModule.to_folder` here to create a link to the API reference

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -293,10 +293,188 @@ Writing Transformations
+``to_folder`` is a method in ``GraphModule`` that allows you to dump
+out the generated FX code to a folder. Although copying the forward pass
+into the code often suffices as in :ref:`Print the generated code
+<the above section>`, it doesn’t capture any model attribute state.
```
Same comment as above, make an anchor label and refer to that

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -293,10 +293,188 @@ Writing Transformations
+To examine modules and parameters, we can use ``to_folder``.
+
+::
+    m = symbolic_trace(M())
```
Newline above

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -293,10 +293,188 @@ Writing Transformations
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+First, read the `FX
+documentation <https://pytorch.org/docs/master/fx.html>`__. Understand
```
This is linking to the page the user is on :). I think it would suffice to `:ref:` to the `Node`, `Graph`, `Proxy` and `Tracer` sections here

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -293,10 +293,188 @@ Writing Transformations
+
+Using the above example, let’s say that the call to ``print(traced)``
+showed us that symbolic tracing had not captured the right code. We’ve
+checked the :ref:`Limitations of Symbolic Tracing`
```
Anchor label

jamesr66a reviewed (commented):
Cool! A bunch of comments about fixing finicky RST rendering

Commit:
Update on "Document FX debugging"


![image](https://user-images.githubusercontent.com/4685384/105525423-f11ca180-5c95-11eb-955d-08e20596b4e0.png)
![image](https://user-images.githubusercontent.com/4685384/105525443-f679ec00-5c95-11eb-8fd3-ee788fa062cd.png)
![image](https://user-images.githubusercontent.com/4685384/105525465-fda0fa00-5c95-11eb-9331-7a47a7e91a11.png)
![image](https://user-images.githubusercontent.com/4685384/105525475-009bea80-5c96-11eb-8e1a-78f8fad2975b.png)



[ghstack-poisoned]

Commit:
Update on "Document FX debugging"


![image](https://user-images.githubusercontent.com/4685384/105525423-f11ca180-5c95-11eb-955d-08e20596b4e0.png)
![image](https://user-images.githubusercontent.com/4685384/105525443-f679ec00-5c95-11eb-8fd3-ee788fa062cd.png)
![image](https://user-images.githubusercontent.com/4685384/105525465-fda0fa00-5c95-11eb-9331-7a47a7e91a11.png)
![image](https://user-images.githubusercontent.com/4685384/105525475-009bea80-5c96-11eb-8e1a-78f8fad2975b.png)



[ghstack-poisoned]

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -321,4 +516,4 @@ API Reference
 .. autoclass:: torch.fx.Tracer
   :members:
 
-.. autoclass:: torch.fx.Proxy
```
Are we losing Proxy here?

Commit:
Update on "Document FX debugging"


![image](https://user-images.githubusercontent.com/4685384/105525423-f11ca180-5c95-11eb-955d-08e20596b4e0.png)
![image](https://user-images.githubusercontent.com/4685384/105525443-f679ec00-5c95-11eb-8fd3-ee788fa062cd.png)
![image](https://user-images.githubusercontent.com/4685384/105525465-fda0fa00-5c95-11eb-9331-7a47a7e91a11.png)
![image](https://user-images.githubusercontent.com/4685384/105525475-009bea80-5c96-11eb-8e1a-78f8fad2975b.png)

Differential Revision: [D26068662](https://our.internmc.facebook.com/intern/diff/D26068662)

[ghstack-poisoned]

Commit:
Update on "Document FX debugging"


![image](https://user-images.githubusercontent.com/4685384/105525423-f11ca180-5c95-11eb-955d-08e20596b4e0.png)
![image](https://user-images.githubusercontent.com/4685384/105525443-f679ec00-5c95-11eb-8fd3-ee788fa062cd.png)
![image](https://user-images.githubusercontent.com/4685384/105525465-fda0fa00-5c95-11eb-9331-7a47a7e91a11.png)
![image](https://user-images.githubusercontent.com/4685384/105525475-009bea80-5c96-11eb-8e1a-78f8fad2975b.png)

Differential Revision: [D26068662](https://our.internmc.facebook.com/intern/diff/D26068662)

[ghstack-poisoned]

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -298,6 +290,209 @@ Miscellanea
+    y = Bar()
+
+After running the above example, we can then look at the code within
+``foo/module.py`` and modify it as desired to debug the generated code.
```
Maybe mention ("(e.g. adding print statements or pdb)") just to make it super obvious that this is the same line of thinking for "how do I add my prints etc"

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -298,6 +290,209 @@ Miscellanea
+Debugging the Transformation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+First, read the FX
```
I think this could use a short transitional sentence such as "Now that we've identified that a transformation is creating incorrect code, we can dig into debugging the transformation itself"

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -298,6 +290,209 @@ Miscellanea
+    """
+    traced.graph.print_tabular()
+
+Using the above example, let’s say that the call to ``print(traced)``
```
This section is titled "Debugging the transformation" but this prose talks about debugging symbolic tracing. I think we should probably focus on the former, as that's where most of the user debugging will happen if we've done our job correctly

jamesr66a reviewed (commented):
Great! Just a few comments

BTW I moved stuff around in `fx.rst` on master so you might run into some merge conflicts during rebase.

Also there is a failure in the docs build:
```
Jan 26 13:38:25   File "/opt/conda/lib/python3.6/site-packages/sphinx/util/inspect.py", line 251, in safe_getattr
Jan 26 13:38:25     return getattr(obj, name, *defargs)
Jan 26 13:38:25 AttributeError: module 'torch.fx' has no attribute 'replace_pattern'
Jan 26 13:38:25 
Jan 26 13:38:25 During handling of the above exception, another exception occurred:
Jan 26 13:38:25 
Jan 26 13:38:25 Traceback (most recent call last):
Jan 26 13:38:25   File "/opt/conda/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
Jan 26 13:38:25     obj = attrgetter(obj, attrname)
Jan 26 13:38:25   File "/opt/conda/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
Jan 26 13:38:25     return autodoc_attrgetter(self.env.app, obj, name, *defargs)
Jan 26 13:38:25   File "/opt/conda/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
Jan 26 13:38:25     return safe_getattr(obj, name, *defargs)
Jan 26 13:38:25   File "/opt/conda/lib/python3.6/site-packages/sphinx/util/inspect.py", line 267, in safe_getattr
Jan 26 13:38:25     raise AttributeError(name)
Jan 26 13:38:25 AttributeError: replace_pattern
```

Commit:
Update on "Document FX debugging"


Differential Revision: [D26068662](https://our.internmc.facebook.com/intern/diff/D26068662)

[ghstack-poisoned]

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -298,6 +290,212 @@ Miscellanea
+read the FX documentation. Once we understand the most important classes
+(:class:`Node`, :class:`Graph`, :class:`Proxy`, and :class:`Tracer`), we
+can determine which class attributes might give us the best
+representation of our program’s intermediate state.
```
This is going the opposite direction of what I suggested. So long as our team does our job right and the `Limitations of Symbolic Tracing` section is sufficient, this section should focus on debugging user's transformations

Commit:
Update on "Document FX debugging"


Differential Revision: [D26068662](https://our.internmc.facebook.com/intern/diff/D26068662)

[ghstack-poisoned]

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -298,6 +290,245 @@ Miscellanea
+Writing Transformations
+-----------------------
+
+TODO
```
I think the rebase got messed up. The order of the sections should be 1) writing transformations 2) debugging transformations 3) limitations of symbolic tracing

jamesr66a commented on torch/fx/__init__.py:
```diff
@@ -82,3 +82,4 @@ def forward(self, x):
 from .graph import Graph
 from .node import Node, map_arg
 from .proxy import Proxy
+from .subgraph_rewriter import replace_pattern
```
Also add this to `fx/__init__.pyi`

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -298,6 +290,245 @@ Miscellanea
+
+    x, y = torch.rand(2, 3), torch.rand(2, 3)
+    m = symbolic_trace(M())
+    import pdb; pdb.set_trace()
```
Might want to add a comment saying something like "When this line is executed at runtime, we will be dropped into an interactive `pdb` prompt. We can use the `step` or `s` command to step into the execution of the next line"

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -298,6 +290,245 @@ Miscellanea
+:ref:`Available Debuggers`.
+
+Debugging the Generated Code
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
The lead-in mentions working backwards, but the examples in this section show directly debugging symbolic tracing. Maybe introduce `my_pass(in : nn.Module) -> nn.Module` and show that getting called, just to keep in theme with working backwards

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -298,6 +290,245 @@ Miscellanea
+:meth:`GraphModule.to_folder` is a method in ``GraphModule`` that allows
+you to dump out the generated FX code to a folder. Although copying the
+forward pass into the code often suffices as in :ref:`Print the Generated Code`,
+it doesn’t capture any model attribute state. To examine modules and parameters,
```
I don't know if "it doesn’t capture any model attribute state" is relevant anymore now that we're creating a subclass in the last example

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -298,6 +290,245 @@ Miscellanea
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Now that we've identified that a transformation is creating incorrect
+code, it's time to debug the transformation itself. We’ve
```
I suggest changing this to something like "First, make sure you've checked the ..."

ansley replied:
I changed it, but I'm not sure how switching between second-person singular and first-person plural sounds, so please take a look

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -298,6 +290,245 @@ Miscellanea
+checked the :ref:`Limitations of Symbolic Tracing` section in the
+documentation, and we know that ``symbolic_trace`` is working as
+expected. Our goal is to figure out what went wrong during our
+transformation of the generated FX ``GraphModule``. First, we need to
```
If the user wrote the transformation, hopefully they're familiar with these classes. Maybe it's better to emphasize reading the docs in the "Writing FX Transformations" section?

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -298,6 +290,245 @@ Miscellanea
+commands <https://docs.python.org/3/library/pdb.html#debugger-commands>`__
+to move through your running program stepwise. It’s common to set a
+breakpoint (``b LINE-NUMBER``) when you start ``pdb`` then call ``c`` to
+run the program until that point. This prevents you from having to step
```
`set_trace()` is used in the above examples so probably a good idea to explicitly mention it

jamesr66a reviewed (commented):
Getting close! Just a few comments. Also I think after both this and the Writing Transformation docs are landed we should take a pass over everything for continuity

Commit:
Update on "Document FX debugging"


Differential Revision: [D26068662](https://our.internmc.facebook.com/intern/diff/D26068662)

[ghstack-poisoned]

jamesr66a reviewed (approved):
Looks good!
